File: php/lib/Builder/Workflow.php
<?php

namespace NativeBPM\Client\Builder;

class Workflow {
    private string $id;
    private string $name;
    private array $nodes = [];
    private array $flows = [];
    private string $currentNodeID = '';
    public array $pendingMerges = [];

    public function __construct(string $id, string $name, $wasmInput = null) {
        $this->id = $id;
        $this->name = $name;
    }

    public function builder(): Workflow {
        return $this;
    }

    private static function toCamelCase(string $s): string {
        if ($s === "wasm") {
            return "wasmPath";
        }
        if ($s === "result_variable") {
            return "resultVar";
        }
        if (!str_contains($s, "_")) {
            return $s;
        }
        $parts = explode('_', $s);
        $camel = $parts[0];
        for ($i = 1; $i < count($parts); $i++) {
            $camel .= ucfirst($parts[$i]);
        }
        if ($camel === "wasm") {
            return "wasmPath";
        }
        if ($camel === "resultVariable") {
            return "resultVar";
        }
        return $camel;
    }

    private static function populateNodeProperties(array &$node, array $opts): void {
        foreach ($opts as $k => $v) {
            $key = self::toCamelCase($k);
            $node[$key] = $v;
        }
    }

    public function startEvent(string $id): Workflow {
        if ($this->findNode($id) !== null) {
            $this->currentNodeID = $id;
            return $this;
        }
        $this->nodes[] = [
            'type' => 'startEvent',
            'id' => $id,
            'name' => 'Start'
        ];
        $this->currentNodeID = $id;
        return $this;
    }

    public function endEvent(string $id, string $name): Workflow {
        if ($this->findNode($id) !== null) {
            return $this;
        }
        $this->nodes[] = [
            'type' => 'endEvent',
            'id' => $id,
            'name' => $name
        ];
        return $this;
    }

    public function serviceTask(string $id, string $name, string $topic, array $options = []): Workflow {
        if ($this->findNode($id) !== null) {
            return $this;
        }
        $node = [
            'type' => 'serviceTask',
            'id' => $id,
            'name' => $name,
            'topic' => $topic
        ];
        self::populateNodeProperties($node, $options);
        $this->nodes[] = $node;
        return $this;
    }

    public function aiTask(string $id, string $name, array $options = []): Workflow {
        if ($this->findNode($id) !== null) {
            return $this;
        }
        $node = [
            'type' => 'aiServiceTask',
            'id' => $id,
            'name' => $name
        ];
        self::populateNodeProperties($node, $options);
        $this->nodes[] = $node;
        return $this;
    }

    public function userTask(string $id, string $name, array $options = []): Workflow {
        if ($this->findNode($id) !== null) {
            return $this;
        }
        $node = [
            'type' => 'userTask',
            'id' => $id,
            'name' => $name
        ];
        self::populateNodeProperties($node, $options);
        $this->nodes[] = $node;
        return $this;
    }

    public function exclusiveGateway(string $id, string $name): Workflow {
        if ($this->findNode($id) !== null) {
            return $this;
        }
        $this->nodes[] = [
            'type' => 'exclusiveGateway',
            'id' => $id,
            'name' => $name
        ];
        return $this;
    }

    public function parallelGateway(string $id, string $name): Workflow {
        if ($this->findNode($id) !== null) {
            return $this;
        }
        $this->nodes[] = [
            'type' => 'parallelGateway',
            'id' => $id,
            'name' => $name
        ];
        return $this;
    }

    public function eventBasedGateway(string $id, string $name): Workflow {
        if ($this->findNode($id) !== null) {
            return $this;
        }
        $this->nodes[] = [
            'type' => 'eventBasedGateway',
            'id' => $id,
            'name' => $name
        ];
        return $this;
    }

    public function callActivity(string $id, string $name, string $calledElement, array $options = []): Workflow {
        if ($this->findNode($id) !== null) {
            return $this;
        }
        $node = [
            'type' => 'callActivity',
            'id' => $id,
            'name' => $name,
            'calledElement' => $calledElement
        ];
        self::populateNodeProperties($node, $options);
        $this->nodes[] = $node;
        return $this;
    }

    public function businessRuleTask(string $id, string $name, string $decisionRef, array $options = []): Workflow {
        if ($this->findNode($id) !== null) {
            return $this;
        }
        $node = [
            'type' => 'businessRuleTask',
            'id' => $id,
            'name' => $name,
            'decisionRef' => $decisionRef
        ];
        self::populateNodeProperties($node, $options);
        $this->nodes[] = $node;
        return $this;
    }

    public function sequenceFlow(string $source, string $target): Workflow {
        $this->flows[] = [
            'id' => "flow-$source-$target",
            'source' => $source,
            'target' => $target,
            'condition' => ''
        ];
        return $this;
    }

    public function sequenceFlowWithCondition(string $source, string $target, string $condition): Workflow {
        $this->flows[] = [
            'id' => "flow-$source-$target",
            'source' => $source,
            'target' => $target,
            'condition' => $condition
        ];
        return $this;
    }

    public function findNode(string $id): ?array {
        foreach ($this->nodes as $key => $node) {
            if ($node['id'] === $id) {
                return $node;
            }
        }
        return null;
    }

    public function updateNode(string $id, array $updatedNode): void {
        foreach ($this->nodes as $key => $node) {
            if ($node['id'] === $id) {
                $this->nodes[$key] = $updatedNode;
                return;
            }
        }
    }

    public function toAST(): array {
        $nodesCopy = $this->nodes;
        $flowsCopy = $this->flows;

        $sourceIds = [];
        foreach ($this->flows as $f) {
            if (isset($f['source'])) {
                $sourceIds[$f['source']] = true;
            }
        }

        foreach ($this->nodes as $node) {
            $nodeType = $node['type'] ?? '';
            $nodeId = $node['id'] ?? '';
            if ($nodeType === 'endEvent' || $nodeType === 'startEvent') {
                continue;
            }
            if (!isset($sourceIds[$nodeId])) {
                $endId = "end_" . $nodeId;
                $nodesCopy[] = [
                    'type' => 'endEvent',
                    'id' => $endId,
                    'name' => 'Process Finished'
                ];
                $flowsCopy[] = [
                    'id' => "flow-" . $nodeId . "-" . $endId,
                    'source' => $nodeId,
                    'target' => $endId,
                    'condition' => ''
                ];
            }
        }

        return [
            'id' => $this->id,
            'name' => $this->name,
            'nodes' => $nodesCopy,
            'flows' => $flowsCopy
        ];
    }

    private function connectNode(string $nodeId): void {
        $node = $this->findNode($nodeId);
        $hasStart = false;
        foreach ($this->nodes as $n) {
            if (($n['type'] ?? '') === 'startEvent') {
                $hasStart = true;
                break;
            }
        }
        if (!$hasStart && $node !== null && ($node['type'] ?? '') !== 'startEvent') {
            $this->startEvent('start');
            $this->sequenceFlow('start', $nodeId);
            $this->currentNodeID = $nodeId;
            return;
        }

        if (count($this->pendingMerges) > 0) {
            foreach ($this->pendingMerges as $sourceId) {
                $this->sequenceFlow($sourceId, $nodeId);
            }
            $this->pendingMerges = [];
        } else if ($this->currentNodeID !== '' && $this->currentNodeID !== $nodeId) {
            $this->sequenceFlow($this->currentNodeID, $nodeId);
        }
        $this->currentNodeID = $nodeId;
    }

    public function start(string $id = 'start'): Workflow {
        $this->startEvent($id);
        $this->connectNode($id);
        return $this;
    }

    public static function V(string $name): Variable {
        return new Variable($name);
    }

    public static function Var(string $name): Variable {
        return new Variable($name);
    }

    public static function v(string $name): Variable {
        return new Variable($name);
    }

    public function end(string $id, string $name): Workflow {
        $this->endEvent($id, $name);
        $this->connectNode($id);
        $this->currentNodeID = '';
        return $this;
    }

    public function user(string $id, string $name, array $options = []): Workflow {
        $this->userTask($id, $name, options);
        $this->connectNode($id);
        return $this;
    }

    public function service(string $id, string $name, string $topic, array $options = []): Workflow {
        $this->serviceTask($id, $name, $topic, options);
        $this->connectNode($id);
        return $this;
    }

    public function ai(string $id, string $name, array $options = []): Workflow {
        $this->aiTask($id, $name, options);
        $this->connectNode($id);
        return $this;
    }

    public function call(string $id, string $name, string $calledElement, array $options = []): Workflow {
        $this->callActivity($id, $name, $calledElement, options);
        $this->connectNode($id);
        return $this;
    }

    public function businessRule(string $id, string $name, string $decisionRef, array $options = []): Workflow {
        $this->businessRuleTask($id, $name, $decisionRef, options);
        $this->connectNode($id);
        return $this;
    }

    public function when($condition): WhenBuilder {
        $gwID = 'gw_' . $this->currentNodeID . '_decision';
        $this->exclusiveGateway($gwID, 'Decision Gateway');
        $this->connectNode($gwID);
        return new WhenBuilder($this, $gwID, (string)$condition);
    }
}

class Branch {
    public Workflow $workflow;
    private string $gatewayID;
    public string $currentNodeID;
    private bool $isConditional;
    private string $condition;
    public bool $hasEnded = false;

    public function __construct(Workflow $workflow, string $gatewayID, string $currentNodeID, bool $isConditional, string $condition) {
        $this->workflow = $workflow;
        $this->gatewayID = $gatewayID;
        $this->currentNodeID = $currentNodeID;
        $this->isConditional = $isConditional;
        $this->condition = $condition;
    }

    private function connectNode(string $nodeId): void {
        if ($this->hasEnded) return;

        $merges = $this->workflow->pendingMerges;
        if (count($merges) > 0) {
            foreach ($merges as $sourceId) {
                $this->workflow->sequenceFlow($sourceId, $nodeId);
            }
            $this->workflow->pendingMerges = [];
            $this->currentNodeID = $nodeId;
            return;
        }

        if ($this->currentNodeID === $this->gatewayID) {
            if ($this->isConditional) {
                $this->workflow->sequenceFlowWithCondition($this->gatewayID, $nodeId, $this->condition);
            } else {
                $this->workflow->sequenceFlow($this->gatewayID, $nodeId);
            }
        } else if ($this->currentNodeID !== '' && $this->currentNodeID !== $nodeId) {
            $this->workflow->sequenceFlow($this->currentNodeID, $nodeId);
        }
        $this->currentNodeID = $nodeId;
    }

    public function user(string $id, string $name, array $options = []): Branch {
        $this->workflow->userTask($id, $name, options);
        $this->connectNode($id);
        return $this;
    }

    public function service(string $id, string $name, string $topic, array $options = []): Branch {
        $this->workflow->serviceTask($id, $name, $topic, options);
        $this->connectNode($id);
        return $this;
    }

    public function ai(string $id, string $name, array $options = []): Branch {
        $this->workflow->aiTask($id, $name, options);
        $this->connectNode($id);
        return $this;
    }

    public function call(string $id, string $name, string $calledElement, array $options = []): Branch {
        $this->workflow->callActivity($id, $name, $calledElement, options);
        $this->connectNode($id);
        return $this;
    }

    public function businessRule(string $id, string $name, string $decisionRef, array $options = []): Branch {
        $this->workflow->businessRuleTask($id, $name, $decisionRef, options);
        $this->connectNode($id);
        return $this;
    }

    public function end(string $id, string $name): Branch {
        $this->workflow->endEvent($id, $name);
        $this->connectNode($id);
        $this->hasEnded = true;
        return $this;
    }

    public function when($condition): WhenBranchBuilder {
        $gwID = 'gw_' . $this->currentNodeID . '_decision';
        $this->workflow->exclusiveGateway($gwID, 'Decision Gateway');
        $this->connectNode($gwID);
        return new WhenBranchBuilder($this, $gwID, (string)$condition);
    }
}

class WhenBuilder {
    private Workflow $workflow;
    private string $gatewayID;
    private string $condition;

    public function __construct(Workflow $workflow, string $gatewayID, string $condition) {
        $this->workflow = $workflow;
        $this->gatewayID = $gatewayID;
        $this->condition = $condition;
    }

    public function then(callable $thenFn): ThenBuilder {
        $thenBranch = new Branch($this->workflow, $this->gatewayID, $this->gatewayID, true, $this->condition);
        $thenFn($thenBranch);

        if (!$thenBranch->hasEnded && $thenBranch->currentNodeID !== $this->gatewayID) {
            $this->workflow->pendingMerges[] = $thenBranch->currentNodeID;
        }

        return new ThenBuilder($this->workflow, $this->gatewayID);
    }
}

class ThenBuilder {
    private Workflow $workflow;
    private string $gatewayID;

    public function __construct(Workflow $workflow, string $gatewayID) {
        $this->workflow = $workflow;
        $this->gatewayID = $gatewayID;
    }

    public function else(callable $elseFn): Workflow {
        $elseBranch = new Branch($this->workflow, $this->gatewayID, $this->gatewayID, false, "");
        $elseFn($elseBranch);

        if (!$elseBranch->hasEnded && $elseBranch->currentNodeID !== $this->gatewayID) {
            $this->workflow->pendingMerges[] = $elseBranch->currentNodeID;
        }

        return $this->workflow;
    }
}

class WhenBranchBuilder {
    private Branch $branch;
    private string $gatewayID;
    private string $condition;

    public function __construct(Branch $branch, string $gatewayID, string $condition) {
        $this->branch = $branch;
        $this->gatewayID = $gatewayID;
        $this->condition = $condition;
    }

    public function then(callable $thenFn): ThenBranchBuilder {
        $thenBranch = new Branch($this->branch->workflow, $this->gatewayID, $this->gatewayID, true, $this->condition);
        $thenFn($thenBranch);

        if (!$thenBranch->hasEnded && $thenBranch->currentNodeID !== $this->gatewayID) {
            $this->branch->workflow->pendingMerges[] = $thenBranch->currentNodeID;
        }

        return new ThenBranchBuilder($this->branch, $this->gatewayID);
    }
}

class ThenBranchBuilder {
    private Branch $branch;
    private string $gatewayID;

    public function __construct(Branch $branch, string $gatewayID) {
        $this->branch = $branch;
        $this->gatewayID = $gatewayID;
    }

    public function else(callable $elseFn): Branch {
        $elseBranch = new Branch($this->branch->workflow, $this->gatewayID, $this->gatewayID, false, "");
        $elseFn($elseBranch);

        if (!$elseBranch->hasEnded && $elseBranch->currentNodeID !== $this->gatewayID) {
            $this->branch->workflow->pendingMerges[] = $elseBranch->currentNodeID;
        }

        return $this->branch;
    }
}

class Expression {
    private string $expr;
    public function __construct(string $expr) {
        $this->expr = $expr;
    }
    public function __toString(): string {
        return $this->expr;
    }
}

class Variable {
    private string $name;
    public function __construct(string $name) {
        $this->name = $name;
    }
    public function eq($val): Expression {
        $valStr = is_bool($val) ? ($val ? 'true' : 'false') : (string)$val;
        return new Expression("{$this->name} == {$valStr}");
    }
    public function ne($val): Expression {
        $valStr = is_bool($val) ? ($val ? 'true' : 'false') : (string)$val;
        return new Expression("{$this->name} != {$valStr}");
    }
    public function gt($val): Expression {
        return new Expression("{$this->name} > {$val}");
    }
    public function gte($val): Expression {
        return new Expression("{$this->name} >= {$val}");
    }
    public function lt($val): Expression {
        return new Expression("{$this->name} < {$val}");
    }
    public function lte($val): Expression {
        return new Expression("{$this->name} <= {$val}");
    }
}

File: java/src/main/java/com/nativebpm/client/builder/Workflow.java
package com.nativebpm.client.builder;

import com.google.gson.Gson;
import java.util.*;
import java.util.function.Consumer;

public class Workflow {
    private final String id;
    private final String name;
    private final List<Map<String, Object>> nodes = new ArrayList<>();
    private final List<Map<String, Object>> flows = new ArrayList<>();

    private String currentNodeID = "";
    private final List<String> pendingMerges = new ArrayList<>();

    public Workflow(String id, String name) {
        this.id = id;
        this.name = name;
    }

    public Workflow builder() {
        return this;
    }

    private static String toCamelCase(String s) {
        if ("wasm".equals(s)) {
            return "wasmPath";
        }
        if ("result_variable".equals(s)) {
            return "resultVar";
        }
        if (!s.contains("_")) {
            return s;
        }
        String[] parts = s.split("_");
        StringBuilder camel = new StringBuilder(parts[0]);
        for (int i = 1; i < parts.length; i++) {
            if (parts[i].length() > 0) {
                camel.append(Character.toUpperCase(parts[i].charAt(0)))
                     .append(parts[i].substring(1));
            }
        }
        String res = camel.toString();
        if ("wasm".equals(res)) {
            return "wasmPath";
        }
        if ("resultVariable".equals(res)) {
            return "resultVar";
        }
        return res;
    }

    private static void populateNodeProperties(Map<String, Object> node, Map<String, Object> opts) {
        if (opts == null) return;
        for (Map.Entry<String, Object> entry : opts.entrySet()) {
            String key = toCamelCase(entry.getKey());
            node.put(key, entry.getValue());
        }
    }

    public static class Branch {
        private final Workflow workflow;
        private final String gatewayID;
        private String currentNodeID;
        private final boolean isConditional;
        private final String condition;
        private boolean hasEnded = false;

        public Branch(Workflow workflow, String gatewayID, String currentNodeID, boolean isConditional, String condition) {
            this.workflow = workflow;
            this.gatewayID = gatewayID;
            this.currentNodeID = currentNodeID;
            this.isConditional = isConditional;
            this.condition = condition;
        }

        private void connectNode(String nodeId) {
            if (hasEnded) return;

            if (!workflow.pendingMerges.isEmpty()) {
                for (String sourceId : workflow.pendingMerges) {
                    workflow.sequenceFlow(sourceId, nodeId);
                }
                workflow.pendingMerges.clear();
                this.currentNodeID = nodeId;
                return;
            }

            if (this.currentNodeID.equals(this.gatewayID)) {
                if (isConditional) {
                    workflow.sequenceFlowWithCondition(this.gatewayID, nodeId, this.condition);
                } else {
                    workflow.sequenceFlow(this.gatewayID, nodeId);
                }
            } else if (!this.currentNodeID.isEmpty() && !this.currentNodeID.equals(nodeId)) {
                workflow.sequenceFlow(this.currentNodeID, nodeId);
            }
            this.currentNodeID = nodeId;
        }

        public Branch user(String id, String name) {
            return user(id, name, null);
        }

        public Branch user(String id, String name, Map<String, Object> options) {
            workflow.userTask(id, name, options);
            connectNode(id);
            return this;
        }

        public Branch service(String id, String name, String topic) {
            return service(id, name, topic, null);
        }

        public Branch service(String id, String name, String topic, Map<String, Object> options) {
            workflow.serviceTask(id, name, topic, options);
            connectNode(id);
            return this;
        }

        public Branch ai(String id, String name) {
            return ai(id, name, null);
        }

        public Branch ai(String id, String name, Map<String, Object> options) {
            workflow.aiTask(id, name, options);
            connectNode(id);
            return this;
        }

        public Branch call(String id, String name, String calledElement) {
            return call(id, name, calledElement, null);
        }

        public Branch call(String id, String name, String calledElement, Map<String, Object> options) {
            workflow.callActivity(id, name, calledElement, options);
            connectNode(id);
            return this;
        }

        public Branch businessRule(String id, String name, String decisionRef) {
            return businessRule(id, name, decisionRef, null);
        }

        public Branch businessRule(String id, String name, String decisionRef, Map<String, Object> options) {
            workflow.businessRuleTask(id, name, decisionRef, options);
            connectNode(id);
            return this;
        }

        public Branch end(String id, String name) {
            workflow.endEvent(id, name);
            connectNode(id);
            this.hasEnded = true;
            return this;
        }

        public WhenBranchBuilder when(String condition) {
            String gwID = "gw_" + this.currentNodeID + "_decision";
            workflow.exclusiveGateway(gwID, "Decision Gateway");
            connectNode(gwID);
            return new WhenBranchBuilder(this, gwID, condition);
        }

        public WhenBranchBuilder when(Expression condition) {
            return when(condition.toString());
        }
    }

    public static class WhenBuilder {
        private final Workflow workflow;
        private final String gatewayID;
        private final String condition;

        public WhenBuilder(Workflow workflow, String gatewayID, String condition) {
            this.workflow = workflow;
            this.gatewayID = gatewayID;
            this.condition = condition;
        }

        public ThenBuilder then(Consumer<Branch> thenFn) {
            Branch thenBranch = new Branch(workflow, gatewayID, gatewayID, true, condition);
            thenFn.accept(thenBranch);

            if (!thenBranch.hasEnded && !thenBranch.currentNodeID.equals(gatewayID)) {
                workflow.pendingMerges.add(thenBranch.currentNodeID);
            }

            return new ThenBuilder(workflow, gatewayID);
        }
    }

    public static class ThenBuilder {
        private final Workflow workflow;
        private final String gatewayID;

        public ThenBuilder(Workflow workflow, String gatewayID) {
            this.workflow = workflow;
            this.gatewayID = gatewayID;
        }

        public Workflow Else(Consumer<Branch> elseFn) {
            Branch elseBranch = new Branch(workflow, gatewayID, gatewayID, false, "");
            elseFn.accept(elseBranch);

            if (!elseBranch.hasEnded && !elseBranch.currentNodeID.equals(gatewayID)) {
                workflow.pendingMerges.add(elseBranch.currentNodeID);
            }

            return workflow;
        }
    }

    public static class WhenBranchBuilder {
        private final Branch branch;
        private final String gatewayID;
        private final String condition;

        public WhenBranchBuilder(Branch branch, String gatewayID, String condition) {
            this.branch = branch;
            this.gatewayID = gatewayID;
            this.condition = condition;
        }

        public ThenBranchBuilder then(Consumer<Branch> thenFn) {
            Branch thenBranch = new Branch(branch.workflow, gatewayID, gatewayID, true, condition);
            thenFn.accept(thenBranch);

            if (!thenBranch.hasEnded && !thenBranch.currentNodeID.equals(gatewayID)) {
                branch.workflow.pendingMerges.add(thenBranch.currentNodeID);
            }

            return new ThenBranchBuilder(branch, gatewayID);
        }
    }

    public static class ThenBranchBuilder {
        private final Branch branch;
        private final String gatewayID;

        public ThenBranchBuilder(Branch branch, String gatewayID) {
            this.branch = branch;
            this.gatewayID = gatewayID;
        }

        public Branch Else(Consumer<Branch> elseFn) {
            Branch elseBranch = new Branch(branch.workflow, gatewayID, gatewayID, false, "");
            elseFn.accept(elseBranch);

            if (!elseBranch.hasEnded && !elseBranch.currentNodeID.equals(gatewayID)) {
                branch.workflow.pendingMerges.add(elseBranch.currentNodeID);
            }

            return branch;
        }
    }

    private void connectNode(String id) {
        Map<String, Object> node = findNode(id);
        boolean hasStart = false;
        for (Map<String, Object> n : nodes) {
            if ("startEvent".equals(n.get("type"))) {
                hasStart = true;
                break;
            }
        }
        if (!hasStart && node != null && !"startEvent".equals(node.get("type"))) {
            startEvent("start");
            sequenceFlow("start", id);
            this.currentNodeID = id;
            return;
        }

        if (!pendingMerges.isEmpty()) {
            for (String sourceID : pendingMerges) {
                sequenceFlow(sourceID, id);
            }
            pendingMerges.clear();
        } else if (!currentNodeID.isEmpty() && !currentNodeID.equals(id)) {
            sequenceFlow(currentNodeID, id);
        }
        currentNodeID = id;
    }

    public Workflow start() {
        return start("start");
    }

    public Workflow start(String id) {
        startEvent(id);
        connectNode(id);
        return this;
    }

    public Workflow end(String id, String name) {
        endEvent(id, name);
        connectNode(id);
        currentNodeID = "";
        return this;
    }

    public Workflow user(String id, String name) {
        return user(id, name, null);
    }

    public Workflow user(String id, String name, Map<String, Object> options) {
        userTask(id, name, options);
        connectNode(id);
        return this;
    }

    public Workflow service(String id, String name, String topic) {
        return service(id, name, topic, null);
    }

    public Workflow service(String id, String name, String topic, Map<String, Object> options) {
        serviceTask(id, name, topic, options);
        connectNode(id);
        return this;
    }

    public Workflow ai(String id, String name) {
        return ai(id, name, null);
    }

    public Workflow ai(String id, String name, Map<String, Object> options) {
        aiTask(id, name, options);
        connectNode(id);
        return this;
    }

    public Workflow call(String id, String name, String calledElement) {
        return call(id, name, calledElement, null);
    }

    public Workflow call(String id, String name, String calledElement, Map<String, Object> options) {
        callActivity(id, name, calledElement, options);
        connectNode(id);
        return this;
    }

    public Workflow businessRule(String id, String name, String decisionRef) {
        return businessRule(id, name, decisionRef, null);
    }

    public Workflow businessRule(String id, String name, String decisionRef, Map<String, Object> options) {
        businessRuleTask(id, name, decisionRef, options);
        connectNode(id);
        return this;
    }

    public WhenBuilder when(String condition) {
        String gwID = "gw_" + this.currentNodeID + "_decision";
        exclusiveGateway(gwID, "Decision Gateway");
        connectNode(gwID);
        return new WhenBuilder(this, gwID, condition);
    }

    public WhenBuilder when(Expression condition) {
        return when(condition.toString());
    }

    public Workflow startEvent(String id) {
        if (findNode(id) != null) {
            this.currentNodeID = id;
            return this;
        }
        Map<String, Object> node = new HashMap<>();
        node.put("type", "startEvent");
        node.put("id", id);
        node.put("name", "Start");
        nodes.add(node);
        this.currentNodeID = id;
        return this;
    }

    public Workflow endEvent(String id, String name) {
        if (findNode(id) != null) return this;
        Map<String, Object> node = new HashMap<>();
        node.put("type", "endEvent");
        node.put("id", id);
        node.put("name", name);
        nodes.add(node);
        return this;
    }

    public Workflow serviceTask(String id, String name, String topic, Map<String, Object> options) {
        if (findNode(id) != null) return this;
        Map<String, Object> node = new HashMap<>();
        node.put("type", "serviceTask");
        node.put("id", id);
        node.put("name", name);
        node.put("topic", topic);
        populateNodeProperties(node, options);
        nodes.add(node);
        return this;
    }

    public Workflow aiTask(String id, String name, Map<String, Object> options) {
        if (findNode(id) != null) return this;
        Map<String, Object> node = new HashMap<>();
        node.put("type", "aiServiceTask");
        node.put("id", id);
        node.put("name", name);
        populateNodeProperties(node, options);
        nodes.add(node);
        return this;
    }

    public Workflow userTask(String id, String name, Map<String, Object> options) {
        if (findNode(id) != null) return this;
        Map<String, Object> node = new HashMap<>();
        node.put("type", "userTask");
        node.put("id", id);
        node.put("name", name);
        populateNodeProperties(node, options);
        nodes.add(node);
        return this;
    }

    public Workflow exclusiveGateway(String id, String name) {
        if (findNode(id) != null) return this;
        Map<String, Object> node = new HashMap<>();
        node.put("type", "exclusiveGateway");
        node.put("id", id);
        node.put("name", name);
        nodes.add(node);
        return this;
    }

    public Workflow parallelGateway(String id, String name) {
        if (findNode(id) != null) return this;
        Map<String, Object> node = new HashMap<>();
        node.put("type", "parallelGateway");
        node.put("id", id);
        node.put("name", name);
        nodes.add(node);
        return this;
    }

    public Workflow eventBasedGateway(String id, String name) {
        if (findNode(id) != null) return this;
        Map<String, Object> node = new HashMap<>();
        node.put("type", "eventBasedGateway");
        node.put("id", id);
        node.put("name", name);
        nodes.add(node);
        return this;
    }

    public Workflow callActivity(String id, String name, String calledElement, Map<String, Object> options) {
        if (findNode(id) != null) return this;
        Map<String, Object> node = new HashMap<>();
        node.put("type", "callActivity");
        node.put("id", id);
        node.put("name", name);
        node.put("calledElement", calledElement);
        populateNodeProperties(node, options);
        nodes.add(node);
        return this;
    }

    public Workflow businessRuleTask(String id, String name, String decisionRef, Map<String, Object> options) {
        if (findNode(id) != null) return this;
        Map<String, Object> node = new HashMap<>();
        node.put("type", "businessRuleTask");
        node.put("id", id);
        node.put("name", name);
        node.put("decisionRef", decisionRef);
        populateNodeProperties(node, options);
        nodes.add(node);
        return this;
    }

    public Workflow sequenceFlow(String source, String target) {
        Map<String, Object> flow = new HashMap<>();
        flow.put("id", "flow-" + source + "-" + target);
        flow.put("source", source);
        flow.put("target", target);
        flow.put("condition", "");
        flows.add(flow);
        return this;
    }

    public Workflow sequenceFlowWithCondition(String source, String target, String condition) {
        Map<String, Object> flow = new HashMap<>();
        flow.put("id", "flow-" + source + "-" + target);
        flow.put("source", source);
        flow.put("target", target);
        flow.put("condition", condition);
        flows.add(flow);
        return this;
    }

    public Map<String, Object> findNode(String id) {
        for (Map<String, Object> node : nodes) {
            if (id.equals(node.get("id"))) {
                return node;
            }
        }
        return null;
    }

    public Map<String, Object> toAST() {
        List<Map<String, Object>> nodesCopy = new ArrayList<>(this.nodes);
        List<Map<String, Object>> flowsCopy = new ArrayList<>(this.flows);

        Set<String> sourceIds = new HashSet<>();
        for (Map<String, Object> f : this.flows) {
            String src = (String) f.get("source");
            if (src != null) {
                sourceIds.add(src);
            }
        }

        for (Map<String, Object> node : this.nodes) {
            String nodeType = (String) node.get("type");
            String nodeId = (String) node.get("id");
            if ("endEvent".equals(nodeType) || "startEvent".equals(nodeType)) {
                continue;
            }
            if (!sourceIds.contains(nodeId)) {
                String endId = "end_" + nodeId;
                
                Map<String, Object> endNode = new HashMap<>();
                endNode.put("type", "endEvent");
                endNode.put("id", endId);
                endNode.put("name", "Process Finished");
                nodesCopy.add(endNode);

                Map<String, Object> flow = new HashMap<>();
                flow.put("id", "flow-" + nodeId + "-" + endId);
                flow.put("source", nodeId);
                flow.put("target", endId);
                flow.put("condition", "");
                flowsCopy.add(flow);
            }
        }

        Map<String, Object> ast = new HashMap<>();
        ast.put("id", this.id);
        ast.put("name", this.name);
        ast.put("nodes", nodesCopy);
        ast.put("flows", flowsCopy);
        return ast;
    }

    public String toJSON() {
        return new Gson().toJson(toAST());
    }

    public static class Expression {
        private final String expr;
        public Expression(String expr) {
            this.expr = expr;
        }
        @Override
        public String toString() {
            return expr;
        }
    }

    public static class Variable {
        private final String name;
        public Variable(String name) {
            this.name = name;
        }
        public Expression eq(Object val) {
            String valStr = (val instanceof Boolean) ? (((Boolean) val) ? "true" : "false") : String.valueOf(val);
            return new Expression(name + " == " + valStr);
        }
        public Expression ne(Object val) {
            String valStr = (val instanceof Boolean) ? (((Boolean) val) ? "true" : "false") : String.valueOf(val);
            return new Expression(name + " != " + valStr);
        }
        public Expression gt(Object val) {
            return new Expression(name + " > " + val);
        }
        public Expression gte(Object val) {
            return new Expression(name + " >= " + val);
        }
        public Expression lt(Object val) {
            return new Expression(name + " < " + val);
        }
        public Expression lte(Object val) {
            return new Expression(name + " <= " + val);
        }
    }

    public static Variable v(String name) {
        return new Variable(name);
    }

    public static Variable V(String name) {
        return new Variable(name);
    }

    public static Variable var(String name) {
        return new Variable(name);
    }
}
